### PR TITLE
Make the spec.pd, spec.tidb, spec.tikv  and spec.discovery, optional in the TidbCluster CR

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve? 
#2971 
### What is changed and how does it work?
Change tikv tidb and discovery type to pointer , and fix  nil  error exception.

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test

Code changes

 - Has Go code change

Side effects

 - None

### CODE REVIEW

**Comment by mikechengwei** [comment]
@DanielZhangQD I think this PR is ok , you can review it now

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -110,16 +110,20 @@ type TidbClusterList struct {
...
 type TidbClusterSpec struct {
 	// Discovery spec
-	Discovery DiscoverySpec `json:"discovery,omitempty"`
+	// +optional
+	Discovery *DiscoverySpec `json:"discovery,omitempty"`
 
 	// PD cluster spec
-	PD PDSpec `json:"pd"`
+	// +optional
+	PD *PDSpec `json:"pd"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:118]:
```suggestion
	PD *PDSpec `json:"pd,omitempty"`
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 2]
```diff
@@ -110,16 +110,20 @@ type TidbClusterList struct {
...
 
 	// TiDB cluster spec
-	TiDB TiDBSpec `json:"tidb"`
+	// +optional
+	TiDB *TiDBSpec `json:"tidb"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:122]:
```suggestion
	TiDB *TiDBSpec `json:"tidb,omitempty"`
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 3]
```diff
@@ -110,16 +110,20 @@ type TidbClusterList struct {
...
 
 	// TiKV cluster spec
-	TiKV TiKVSpec `json:"tikv"`
+	// +optional
+	TiKV *TiKVSpec `json:"tikv"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:126]:
```suggestion
	TiKV *TiKVSpec `json:"tikv,omitempty"`
```

File: pkg/manager/member/tidb_discovery_manager.go [hunk 4]
```diff
@@ -47,6 +47,10 @@ func NewTidbDiscoveryManager(typedControl controller.TypedControlInterface) Tidb
 
 func (m *realTidbDiscoveryManager) Reconcile(tc *v1alpha1.TidbCluster) error {
 
+	// If discovery is not specified return
+	if tc.Spec.Discovery == nil {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_discovery_manager.go:51]: Currently, the `Spec.Discovery` only specifies the resource requirement of discovery component, actually, the creation of discovery is not controlled by the `Spec.Discovery`, I think we can check the `spec.pd ==nil` here for this PR.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 5]
```diff
@@ -1329,4 +1329,5 @@ func newTidbClusterConfig(cfg *tests.Config, ns, clusterName, password, tidbVers
 		EnableConfigMapRollout: true,
 		ClusterVersion:         tidbVersion,
 	}
+
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1332]: No need to change this.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 6]
```diff
@@ -110,16 +110,20 @@ type TidbClusterList struct {
 // TidbClusterSpec describes the attributes that a user creates on a tidb cluster
 type TidbClusterSpec struct {
 	// Discovery spec
-	Discovery DiscoverySpec `json:"discovery,omitempty"`
+	// +optional
+	Discovery *DiscoverySpec `json:"discovery,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:114]: No need to update this as there is only resource configurations in this field.

Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/types.go]: ok

File: examples/prometheus-operator/tidb-cluster.yaml [hunk 7]
```diff
@@ -6,6 +6,7 @@ spec:
   version: v3.0.8
   timezone: UTC
   pvReclaimPolicy: Delete
+  discovery: {}
```
Comment (DanielZhangQD) [inline examples/prometheus-operator/tidb-cluster.yaml:9]: No need to change for the `discovery`.

File: examples/tiflash/tidb-cluster.yaml [hunk 8]
```diff
@@ -6,6 +6,7 @@ spec:
   configUpdateStrategy: RollingUpdate
   enablePVReclaim: false
   imagePullPolicy: IfNotPresent
+  discovery: {}
```
Comment (DanielZhangQD) [inline examples/tiflash/tidb-cluster.yaml:9]: ditto

File: pkg/apis/pingcap/v1alpha1/tidbcluster.go [hunk 9]
```diff
@@ -438,6 +438,9 @@ func (tc *TidbCluster) TiDBStsDesiredOrdinals(excludeFailover bool) sets.Int32 {
 }
 
 func (tc *TidbCluster) PDIsAvailable() bool {
+	if tc.Spec.PD == nil {
+		return false
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbcluster.go:442]: We may have to return `true` here, or this may block the upgrade of other components.

**Pull request merged**